File: android/app/src/main/kotlin/com/vyooo/insta360/pipeline/HeuristicDecisionLayer.kt
package com.vyooo.insta360.pipeline

import kotlin.math.abs

/**
 * Milestone 3 — bounded, on-device, **heuristic** AI-assisted decision layer (the "AI Assist" /
 * "On-device AI" module in the patent diagrams).
 *
 * It observes low-cost scene statistics from a coarse luma grid (the "preview / luma" input) and
 * writes **decision signals only** into the shared [MutableHints] that the deterministic pipeline
 * already consumes. It never sees or modifies pixel content beyond reading a downscaled luma sample,
 * and the pipeline runs unchanged when this layer is disabled (deterministic fall-open).
 *
 * ## Contract with the temporal pacer (M2, time-based variant)
 * [TemporalDedupStage] is a **time-based motion/static pacer**: while *moving* it keeps every frame
 * (full rate); while *static* it keeps one frame per `1/staticFps` s. Its gate is a **binary**
 * compare, `motion >= motionThreshold`. Three consequences shape this layer:
 *
 *  1. **We emit a decision, not a raw metric.** [MutableHints.motion] is set to [SIGNAL_MOVING] /
 *     [SIGNAL_STATIC] (1 / 0), so the gate resolves the same way regardless of how the stage's
 *     threshold is tuned. Raw activity stays observable via [stats] (`aiActivity`). Emitting a raw
 *     top-percentile score instead would be *miscalibrated*: the stage's threshold (0.010) was tuned
 *     against a whole-frame **mean**, while a top-percentile score sits far above it even for sensor
 *     noise — a static scene would read as "moving" and the reduction would silently never engage.
 *  2. **The gate must not flap.** An unstable moving/static decision produces *irregular* frame
 *     spacing — the exact failure that made the earlier frame-count scheduler lag the encoder. So the
 *     decision uses **hysteresis** ([motionEnter] > [motionExit]) plus a **hold** ([motionHoldMs])
 *     after activity subsides.
 *  3. **Slow motion must accumulate.** The stage's own metric compares against the last *kept* frame,
 *     so a slow pan builds up and is not silently dropped. Comparing consecutive frames would lose
 *     that. We reproduce it by **freezing the reference grid while static** (drift accumulates against
 *     it) and refreshing it while moving.
 *
 * ## Signals (contract §3.4 / Patent Fig. 2–3)
 *  - **motion** → [TemporalDedupStage]: hysteretic moving/static decision from a robust,
 *    localized-motion-aware activity score (top-percentile luma difference, EMA-smoothed). Unlike a
 *    whole-frame mean, a small moving region is not averaged away. **Applied live when [enabled].**
 *  - **perceptualScale** → [DownscaleStage]: scene-complexity score → a recommended reduction.
 *    Computed + exposed; applied to the hint only when [applyPerceptual] (default off, for a stable
 *    live resolution — demonstration-grade per the contract).
 *  - **forwardThetaDeg** → [ForwardMaskStage]: salience-centroid orientation, heavily smoothed.
 *    Computed + exposed; applied only when [applyOrientation] (default off — the mask stays put; the
 *    field size F is never changed, per the patent).
 *
 * Constraints honoured: metadata-only, no pixel modification, no training/cloud, deterministic
 * fall-open, bounded (a few hundred luma reads per observation).
 *
 * **[enabled] defaults to `false`**: the layer ships inert so the live path keeps the temporal
 * behaviour already validated on-device. Turn it on deliberately for A/B / KPI capture via
 * `Insta360FrameSink.setAiEnabled(true)`.
 *
 * State is touched only on the SDK extract thread (via `Insta360FrameSink.submit`); the toggles are
 * `@Volatile` for live control from another thread.
 */
class HeuristicDecisionLayer(private val hints: MutableHints) {

    /**
     * Master enable. **On for this integration/test branch** so M3 is exercised on-device without a UI
     * toggle (nothing in the app calls `setAiEnabled` yet). It ships **off** on the M3 branch, where the
     * pipeline falls open to its own deterministic metrics (the validated M2 behaviour). A/B at runtime
     * via `Insta360FrameSink.setAiEnabled(false)`.
     */
    @Volatile var enabled: Boolean = true

    /** Apply the computed spatial-reduction recommendation to the hint (default off = mapping only). */
    @Volatile var applyPerceptual: Boolean = false

    /** Apply the computed forward orientation to the hint (default off = mapping only). */
    @Volatile var applyOrientation: Boolean = false

    /** Observe every Nth frame (≥1). 1 = every frame (coarse-grid cost is sub-millisecond). */
    @Volatile var cadence: Int = 1

    // ── Motion decision tuning (top-percentile activity scale, NOT the stage's mean scale) ───
    /** Activity at/above which a static scene is declared **moving**. */
    @Volatile var motionEnter: Float = 0.030f

    /** Activity below which a moving scene may fall back to **static** (hysteresis: < [motionEnter]). */
    @Volatile var motionExit: Float = 0.015f

    /** Stay "moving" this long after activity drops below [motionExit] — damps gate flapping. */
    @Volatile var motionHoldMs: Long = 400

    // ── Runtime state (extract thread) ───────────────────────────────────────
    /** Motion-diff reference: refreshed while moving, **frozen while static** so slow drift accumulates. */
    private var refGrid: IntArray? = null
    private var frameCounter = 0L
    private var activityEma = 0f
    private var detailEma = 0f
    private var thetaEma = 0f
    private var moving = false
    private var belowExitSinceNs = 0L

    // ── Exposed signals + performance metrics ────────────────────────────────
    @Volatile private var lastActivity = 0f
    @Volatile private var lastMoving = false
    @Volatile private var lastPerceptualScale = 1f
    @Volatile private var lastRecommendedScale = 1f
    @Volatile private var lastThetaDeg = 0f
    @Volatile private var lastDetail = 0f
    @Volatile private var decisions = 0L
    @Volatile private var motionSpans = 0L
    @Volatile private var overheadMsEma = 0.0

    /** Observe one frame's luma and update the decision signals. Cheap; safe to call per frame. */
    fun observe(pixels: ByteArray, width: Int, height: Int) {
        if (!enabled) {
            clearHints()
            return
        }
        frameCounter++
        val n = cadence.coerceAtLeast(1)
        // Between observations the previous decision stands (the hint is sticky) — never left stale-null.
        if (frameCounter % n != 0L) return

        val t0 = System.nanoTime()
        val grid = sampleLuma(pixels, width, height) ?: return

        // Activity vs the reference grid. The reference is frozen while static, so a slow pan
        // accumulates against it and is eventually detected (mirrors the stage's last-kept baseline).
        // Top-percentile (not whole-frame mean) keeps a small moving region from being averaged away.
        val ref = refGrid
        val activity = if (ref == null) 1f else topPercentileDiff(grid, ref)
        activityEma = MOTION_EMA_ALPHA * activity + (1 - MOTION_EMA_ALPHA) * activityEma

        // Hysteretic moving/static decision + hold. A stable gate is essential: a flapping decision
        // gives the encoder irregular frame spacing, which is what lagged the earlier scheduler.
        if (moving) {
            if (activityEma < motionExit) {
                if (belowExitSinceNs == 0L) belowExitSinceNs = t0
                if ((t0 - belowExitSinceNs) / 1_000_000L >= motionHoldMs) {
                    moving = false
                    belowExitSinceNs = 0L
                }
            } else {
                belowExitSinceNs = 0L // re-armed by fresh activity
            }
        } else if (activityEma >= motionEnter) {
            moving = true
            belowExitSinceNs = 0L
            motionSpans++
        }

        // Refresh the reference only while moving; freeze it while static (see above).
        if (moving || ref == null) refGrid = grid

        // Spatial detail: mean neighbour gradient of the grid (texture/complexity estimate).
        val detail = neighbourGradient(grid)
        detailEma = DETAIL_EMA_ALPHA * detail + (1 - DETAIL_EMA_ALPHA) * detailEma

        // Perceptual score in [0,1] (1 = keep full resolution). More detail/activity → keep more.
        val perceptual = (0.5f + 2.0f * detailEma + 0.5f * activityEma).coerceIn(0f, 1f)
        // Recommended reduction factor of the 2K target (bounded — never below RECOMMEND_FLOOR).
        val recommended = (RECOMMEND_FLOOR + (1f - RECOMMEND_FLOOR) * perceptual).coerceIn(RECOMMEND_FLOOR, 1f)

        // Forward orientation: salience-weighted centroid column → θ in degrees, heavily smoothed.
        val theta = salienceCentroidDeg(grid)
        thetaEma = THETA_EMA_ALPHA * theta + (1 - THETA_EMA_ALPHA) * thetaEma

        // Write decision signals (metadata only). `motion` is a DECISION (moving/static), not a raw
        // score — the stage's binary gate then resolves identically whatever its threshold is set to.
        hints.motion = if (moving) SIGNAL_MOVING else SIGNAL_STATIC
        hints.perceptualScale = if (applyPerceptual) perceptual else null
        hints.forwardThetaDeg = if (applyOrientation) thetaEma else null
        // isPanoramic left to the deterministic detector (corroboration hook, not overridden).

        lastActivity = activityEma
        lastMoving = moving
        lastPerceptualScale = perceptual
        lastRecommendedScale = recommended
        lastThetaDeg = thetaEma
        lastDetail = detailEma
        decisions++
        val ms = (System.nanoTime() - t0) / 1_000_000.0
        overheadMsEma = if (decisions == 1L) ms else 0.1 * ms + 0.9 * overheadMsEma
    }

    /** Performance-observation + influence-mapping metrics (M3 deliverable #4). */
    fun stats(): Map<String, Any> = mapOf(
        "aiEnabled" to enabled,
        "aiApplyPerceptual" to applyPerceptual,
        "aiApplyOrientation" to applyOrientation,
        "aiDecisions" to decisions,
        "aiOverheadMs" to overheadMsEma,
        "aiMoving" to lastMoving,          // the decision handed to the temporal pacer
        "aiActivity" to lastActivity,      // raw smoothed activity behind that decision
        "aiMotionSpans" to motionSpans,    // static→moving transitions (gate-stability KPI)
        "aiSpatialDetail" to lastDetail,
        "aiPerceptualScale" to lastPerceptualScale,
        "aiRecommendedScale" to lastRecommendedScale,
        "aiThetaDeg" to lastThetaDeg,
    )

    fun reset() {
        refGrid = null
        frameCounter = 0
        activityEma = 0f; detailEma = 0f; thetaEma = 0f
        moving = false; belowExitSinceNs = 0L
        lastActivity = 0f; lastMoving = false
        lastPerceptualScale = 1f; lastRecommendedScale = 1f; lastThetaDeg = 0f; lastDetail = 0f
        decisions = 0; motionSpans = 0; overheadMsEma = 0.0
        clearHints()
    }

    private fun clearHints() {
        hints.motion = null
        hints.perceptualScale = null
        hints.forwardThetaDeg = null
    }

    // ── Heuristics ───────────────────────────────────────────────────────────

    /** Mean of the top [TOP_FRACTION] cell differences, normalised to [0,1] — localized-motion aware. */
    private fun topPercentileDiff(cur: IntArray, prev: IntArray): Float {
        if (cur.size != prev.size || cur.isEmpty()) return 1f
        val diffs = IntArray(cur.size) { abs(cur[it] - prev[it]) }
        diffs.sort() // ascending
        val k = (cur.size * TOP_FRACTION).toInt().coerceAtLeast(1)
        var acc = 0L
        for (i in cur.size - k until cur.size) acc += diffs[i]
        return (acc.toDouble() / (k * 255.0)).toFloat().coerceIn(0f, 1f)
    }

    /** Mean absolute right/below neighbour gradient of the grid, normalised to [0,1]. */
    private fun neighbourGradient(grid: IntArray): Float {
        var acc = 0L
        var count = 0
        for (r in 0 until GRID_ROWS) {
            for (c in 0 until GRID_COLS) {
                val v = grid[r * GRID_COLS + c]
                if (c + 1 < GRID_COLS) { acc += abs(v - grid[r * GRID_COLS + c + 1]); count++ }
                if (r + 1 < GRID_ROWS) { acc += abs(v - grid[(r + 1) * GRID_COLS + c]); count++ }
            }
        }
        if (count == 0) return 0f
        return (acc.toDouble() / (count * 255.0)).toFloat().coerceIn(0f, 1f)
    }

    /** Salience-weighted centroid column → θ in degrees ([-180,180], 0 = ERP centre / forward). */
    private fun salienceCentroidDeg(grid: IntArray): Float {
        var wsum = 0.0
        var vsum = 0.0
        for (c in 0 until GRID_COLS) {
            var colSal = 0L
            for (r in 0 until GRID_ROWS) colSal += grid[r * GRID_COLS + c]
            wsum += c.toDouble() * colSal
            vsum += colSal.toDouble()
        }
        if (vsum <= 0.0) return 0f
        val centroidCol = wsum / vsum
        return (((centroidCol / GRID_COLS) - 0.5) * 360.0).toFloat()
    }

    /** Coarse luma grid (centre pixel per cell). Null if the frame buffer is unusable. */
    private fun sampleLuma(px: ByteArray, w: Int, h: Int): IntArray? {
        if (w <= 0 || h <= 0 || px.size < w * h * 4) return null
        val out = IntArray(GRID_COLS * GRID_ROWS)
        var k = 0
        for (ry in 0 until GRID_ROWS) {
            val y = ((ry + 0.5f) * h / GRID_ROWS).toInt().coerceIn(0, h - 1)
            val rowBase = y * w
            for (cx in 0 until GRID_COLS) {
                val x = ((cx + 0.5f) * w / GRID_COLS).toInt().coerceIn(0, w - 1)
                val idx = (rowBase + x) * 4
                val r = px[idx].toInt() and 0xFF
                val g = px[idx + 1].toInt() and 0xFF
                val b = px[idx + 2].toInt() and 0xFF
                out[k++] = (r * 77 + g * 150 + b * 29) shr 8 // BT.601 luma
            }
        }
        return out
    }

    private companion object {
        const val GRID_COLS = 32
        const val GRID_ROWS = 18
        const val TOP_FRACTION = 0.15f
        const val MOTION_EMA_ALPHA = 0.4f
        const val DETAIL_EMA_ALPHA = 0.2f
        const val THETA_EMA_ALPHA = 0.05f // heavy smoothing → stable forward direction
        const val RECOMMEND_FLOOR = 0.5f // never recommend below 50% of the 2K target

        /** Decision values written to [MutableHints.motion] — read by the pacer's binary gate. */
        const val SIGNAL_MOVING = 1.0f
        const val SIGNAL_STATIC = 0.0f
    }
}

File: ios/Runner/Insta360/Pipeline/HeuristicDecisionLayer.swift
import Foundation

/// Milestone 3 — bounded, on-device, heuristic AI-assisted decision layer. Swift port of Android
/// `HeuristicDecisionLayer.kt` (every constant and rule mirrored exactly).
///
/// Observes low-cost scene statistics from a coarse luma grid and writes decision signals only into
/// the shared `MutableHints` the deterministic pipeline consumes. It never modifies pixel content and
/// the pipeline runs unchanged when disabled (deterministic fall-open).
///
/// Emits: `motion` (hysteretic moving/static decision → `TemporalDedupStage`), `perceptualScale`
/// (scene-complexity → `DownscaleStage`, applied when `applyPerceptual`), `forwardThetaDeg`
/// (salience-centroid orientation, applied when `applyOrientation`).
final class HeuristicDecisionLayer {

    private let hints: MutableHints

    init(hints: MutableHints) { self.hints = hints }

    /// Master enable. On → supplies the decision signals the pipeline acts on.
    var enabled = true
    /// Emit the computed spatial-reduction recommendation as a live hint (M3 spatial influence).
    var applyPerceptual = true
    /// Apply the computed forward orientation to the hint (default off = mapping only).
    var applyOrientation = false
    /// Observe every Nth frame (≥1).
    var cadence = 1

    // ── Motion decision tuning ───────────────────────────────────────────────
    var motionEnter: Float = 0.030
    var motionExit: Float = 0.015
    var motionHoldMs: Int64 = 400
    var refLagMs: Int64 = 250

    // ── Runtime state (extract thread) ───────────────────────────────────────
    private var history: [(UInt64, [Int])] = []
    private var frameCounter: Int64 = 0
    private var activityEma: Float = 0
    private var detailEma: Float = 0
    private var thetaEma: Float = 0
    private var moving = false
    private var belowExitSinceNs: UInt64 = 0

    // ── Exposed signals + performance metrics ────────────────────────────────
    private var lastActivity: Float = 0
    private var lastMoving = false
    private var lastPerceptualScale: Float = 1
    private var lastRecommendedScale: Float = 1
    private var lastThetaDeg: Float = 0
    private var lastDetail: Float = 0
    private var decisions: Int64 = 0
    private var motionSpans: Int64 = 0
    private var overheadMsEma: Double = 0

    /// Observe one frame's luma and update the decision signals.
    func observe(_ pixels: [UInt8], width: Int, height: Int) {
        if !enabled {
            clearHints()
            return
        }
        frameCounter += 1
        let n = max(cadence, 1)
        // Between observations the previous decision stands (the hint is sticky).
        if frameCounter % Int64(n) != 0 { return }

        let t0 = DispatchTime.now().uptimeNanoseconds
        guard let grid = sampleLuma(pixels, width, height) else { return }

        // Activity vs a FIXED ~refLagMs time-lagged reference — never a state-dependent one.
        history.append((t0, grid))
        let lagNs = UInt64(refLagMs) * 1_000_000
        while history.count > 1 && t0 - history[0].0 > lagNs { history.removeFirst() }
        let ref = history.count > 1 ? history[0].1 : nil
        let activity = ref == nil ? Float(1) : topPercentileDiff(grid, ref!)
        activityEma = Self.motionEmaAlpha * activity + (1 - Self.motionEmaAlpha) * activityEma

        // Hysteretic moving/static decision + hold.
        if moving {
            if activityEma < motionExit {
                if belowExitSinceNs == 0 { belowExitSinceNs = t0 }
                if Int64((t0 - belowExitSinceNs) / 1_000_000) >= motionHoldMs {
                    moving = false
                    belowExitSinceNs = 0
                    NSLog("[HeuristicAI] gate moving->STATIC activity=\(activityEma) exit=\(motionExit) hold=\(motionHoldMs)ms")
                }
            } else {
                belowExitSinceNs = 0 // re-armed by fresh activity
            }
        } else if activityEma >= motionEnter {
            moving = true
            belowExitSinceNs = 0
            motionSpans += 1
            NSLog("[HeuristicAI] gate static->MOVING activity=\(activityEma) enter=\(motionEnter) spans=\(motionSpans)")
        }

        // Spatial detail: mean neighbour gradient of the grid.
        let detail = neighbourGradient(grid)
        detailEma = Self.detailEmaAlpha * detail + (1 - Self.detailEmaAlpha) * detailEma

        // Perceptual score in [0,1] (1 = keep full resolution).
        let perceptual = min(max(0.5 + 2.0 * detailEma + 0.5 * activityEma, 0), 1)
        // Recommended reduction factor of the 2K target (never below RECOMMEND_FLOOR).
        let recommended = min(max(Self.recommendFloor + (1 - Self.recommendFloor) * perceptual,
                                  Self.recommendFloor), 1)

        // Forward orientation: salience-weighted centroid column → θ in degrees, heavily smoothed.
        let theta = salienceCentroidDeg(grid)
        thetaEma = Self.thetaEmaAlpha * theta + (1 - Self.thetaEmaAlpha) * thetaEma

        // Write decision signals (metadata only). `motion` is a DECISION (moving/static), not a score.
        hints.motion = moving ? Self.signalMoving : Self.signalStatic
        hints.perceptualScale = applyPerceptual ? perceptual : nil
        hints.forwardThetaDeg = applyOrientation ? thetaEma : nil
        // isPanoramic left to the deterministic detector.

        lastActivity = activityEma
        lastMoving = moving
        lastPerceptualScale = perceptual
        lastRecommendedScale = recommended
        lastThetaDeg = thetaEma
        lastDetail = detailEma
        decisions += 1
        let ms = Double(DispatchTime.now().uptimeNanoseconds - t0) / 1_000_000.0
        overheadMsEma = decisions == 1 ? ms : 0.1 * ms + 0.9 * overheadMsEma
    }

    /// Performance-observation + influence-mapping metrics (M3 deliverable #4).
    func stats() -> [String: Any] {
        [
            "aiEnabled": enabled,
            "aiApplyPerceptual": applyPerceptual,
            "aiApplyOrientation": applyOrientation,
            "aiDecisions": decisions,
            "aiOverheadMs": overheadMsEma,
            "aiMoving": lastMoving,
            "aiActivity": lastActivity,
            "aiMotionSpans": motionSpans,
            "aiSpatialDetail": lastDetail,
            "aiPerceptualScale": lastPerceptualScale,
            "aiRecommendedScale": lastRecommendedScale,
            "aiThetaDeg": lastThetaDeg,
        ]
    }

    func reset() {
        history.removeAll()
        frameCounter = 0
        activityEma = 0; detailEma = 0; thetaEma = 0
        moving = false; belowExitSinceNs = 0
        lastActivity = 0; lastMoving = false
        lastPerceptualScale = 1; lastRecommendedScale = 1; lastThetaDeg = 0; lastDetail = 0
        decisions = 0; motionSpans = 0; overheadMsEma = 0
        clearHints()
    }

    private func clearHints() {
        hints.motion = nil
        hints.perceptualScale = nil
        hints.forwardThetaDeg = nil
    }

    // ── Heuristics ───────────────────────────────────────────────────────────

    /// Mean of the top `TOP_FRACTION` cell differences, normalised to [0,1] — localized-motion aware.
    private func topPercentileDiff(_ cur: [Int], _ prev: [Int]) -> Float {
        if cur.count != prev.count || cur.isEmpty { return 1 }
        var diffs = [Int](repeating: 0, count: cur.count)
        for i in cur.indices { diffs[i] = abs(cur[i] - prev[i]) }
        diffs.sort() // ascending
        let k = max(Int(Float(cur.count) * Self.topFraction), 1)
        var acc: Int64 = 0
        for i in (cur.count - k)..<cur.count { acc += Int64(diffs[i]) }
        return Float(min(max(Double(acc) / (Double(k) * 255.0), 0), 1))
    }

    /// Mean absolute right/below neighbour gradient of the grid, normalised to [0,1].
    private func neighbourGradient(_ grid: [Int]) -> Float {
        var acc: Int64 = 0
        var count = 0
        for r in 0..<Self.gridRows {
            for c in 0..<Self.gridCols {
                let v = grid[r * Self.gridCols + c]
                if c + 1 < Self.gridCols { acc += Int64(abs(v - grid[r * Self.gridCols + c + 1])); count += 1 }
                if r + 1 < Self.gridRows { acc += Int64(abs(v - grid[(r + 1) * Self.gridCols + c])); count += 1 }
            }
        }
        if count == 0 { return 0 }
        return Float(min(max(Double(acc) / (Double(count) * 255.0), 0), 1))
    }

    /// Salience-weighted centroid column → θ in degrees ([-180,180], 0 = ERP centre / forward).
    private func salienceCentroidDeg(_ grid: [Int]) -> Float {
        var wsum = 0.0
        var vsum = 0.0
        for c in 0..<Self.gridCols {
            var colSal: Int64 = 0
            for r in 0..<Self.gridRows { colSal += Int64(grid[r * Self.gridCols + c]) }
            wsum += Double(c) * Double(colSal)
            vsum += Double(colSal)
        }
        if vsum <= 0 { return 0 }
        let centroidCol = wsum / vsum
        return Float(((centroidCol / Double(Self.gridCols)) - 0.5) * 360.0)
    }

    /// Coarse luma grid (centre pixel per cell). Nil if the frame buffer is unusable.
    private func sampleLuma(_ px: [UInt8], _ w: Int, _ h: Int) -> [Int]? {
        if w <= 0 || h <= 0 || px.count < w * h * 4 { return nil }
        var out = [Int](repeating: 0, count: Self.gridCols * Self.gridRows)
        px.withUnsafeBufferPointer { buf in
            guard let p = buf.baseAddress else { return }
            var k = 0
            for ry in 0..<Self.gridRows {
                let y = min(max(Int((Float(ry) + 0.5) * Float(h) / Float(Self.gridRows)), 0), h - 1)
                let rowBase = y * w
                for cx in 0..<Self.gridCols {
                    let x = min(max(Int((Float(cx) + 0.5) * Float(w) / Float(Self.gridCols)), 0), w - 1)
                    let idx = (rowBase + x) * 4
                    let r = Int(p[idx]); let g = Int(p[idx + 1]); let b = Int(p[idx + 2])
                    out[k] = (r * 77 + g * 150 + b * 29) >> 8 // BT.601 luma
                    k += 1
                }
            }
        }
        return out
    }

    // MARK: - Constants (mirrored exactly from Android)
    private static let gridCols = 32
    private static let gridRows = 18
    private static let topFraction: Float = 0.15
    private static let motionEmaAlpha: Float = 0.4
    private static let detailEmaAlpha: Float = 0.2
    private static let thetaEmaAlpha: Float = 0.05
    private static let recommendFloor: Float = 0.5
    private static let signalMoving: Float = 1.0
    private static let signalStatic: Float = 0.0
}
